token-wrap/program/src/processor.rs

```diff
@@ -1,26 +1,915 @@
+                wrapped_mint.key,
+                user_destination_account.key,
+                 signer.key,
+                &[],
+                amount,
+                unpacked_mint.decimals,
+            )?;
+            invoke_signed(
+                &mint_to_user_account_ix,
+                &[
+                    wrapped_mint.clone(),
+                    user_destination_account.clone(),
+                    wrapped_token_program.clone(),
+                ],
+                &[wrapped_mint_authority_seeds],
+            )?;
+        },
+        _ => {
+            // Handle unknown or unsupported token program
+            return Err(ProgramError::InvalidAccountData);
+        },
+    }
+
+    Ok(())
 }
+/// Process 'Unwrap' instruction
+fn process_unwrap(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    input: &[u8],
+) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+    msg!("123");
+    
+    let wrapped_token_account = next_account_info(account_info_iter)?;
+    let wrapped_mint = next_account_info(account_info_iter)?;
+    let escrow_account = next_account_info(account_info_iter)?;
+    let user_unwrapped_token_account = next_account_info(account_info_iter)?;
+    let unwrapped_mint = next_account_info(account_info_iter)?;
+    let escrow_authority_account = next_account_info(account_info_iter)?;
+    let wrapped_token_program = next_account_info(account_info_iter)?;
+    let unwrapped_token_program = next_account_info(account_info_iter)?;
+    let signer = next_account_info(account_info_iter)?;
+
+
+// Parse the amount to unwrap from the input
+let amount = unpack_amount(input[1..9].try_into().unwrap())?;
+
+// Derive the seeds and bump for the wrapped mint authority
+let (wrapped_mint_authority, bump_seed) = get_wrapped_mint_authority_with_seed(wrapped_token_account.key);
+let wrapped_mint_authority_seeds = get_wrapped_mint_authority_seeds(wrapped_token_account.key);
+let signer_seeds = &[
+    &wrapped_mint_authority_seeds[0][..],
+    &wrapped_mint_authority_seeds[1][..],
+    &[bump_seed],
+];
+
+
+assert!(wrapped_token_account.is_signer, "{}", ProgramError::MissingRequiredSignature);
+assert_eq!(wrapped_token_account.owner, wrapped_token_program.key, "{}", ProgramError::IncorrectProgramId);
+assert!(amount > 0, "{}", ProgramError::InvalidArgument);
+assert!(escrow_account.lamports() >= amount, "{}", ProgramError::InsufficientFunds);
+assert_eq!(user_unwrapped_token_account.owner, unwrapped_token_program.key, "{}", ProgramError::IncorrectProgramId);
+assert!(signer.is_signer, "{}", ProgramError::MissingRequiredSignature);
+
+
+       
+let (_, bump_seed) = get_wrapped_mint_address_with_seed(unwrapped_mint.key, wrapped_token_program.key);
+let bumps = &[bump_seed];
+let wrapped_mint_authority_seeds: &[&[u8]]  = &get_wrapped_mint_signer_seeds(unwrapped_mint.key, wrapped_token_program.key, bumps);
+
+match *wrapped_token_program.key {
+    spl_token::ID => {
```

Comment:
No need to change the logic based on the token program! This should work with any number of token programs that conform to the interface